devito/types/array.py

```diff
@@ -345,3 +346,217 @@ def dim(self):
+
+    @classmethod
+    def __args_setup__(cls, *args, **kwargs):
+        components = kwargs.get('components', ())
+        if len(components) <= 1:
+            raise ValueError("Expected at least two components")
+
+        klss = {type(i).__base__ for i in components}
+        if len(klss) != 1:
+            raise ValueError("Components must be of same type")
+        if not issubclass(klss.pop(), AbstractFunction):
+            raise ValueError("Component type must be subclass of AbstractFunction")
+
+        return args, kwargs
+
+    @classmethod
+    def __dtype_setup__(cls, components=(), **kwargs):
+        dtypes = {i.dtype for i in components}
+        if len(dtypes) > 1:
+            raise ValueError("Components must have the same dtype")
+        dtype = dtypes.pop()
+        try:
+            return dtypes_vector_mapper[(dtype, len(components))]
+        except KeyError:
+            raise NotImplementedError("Unsupported vector type `%s`" % dtype)
+
+    @classmethod
+    def __indices_setup__(cls, components=(), **kwargs):
+        dimensionss = {i.dimensions for i in components}
+        if len(dimensionss) > 1:
+            raise ValueError("Components must have the same dimensions")
+        dimensions = dimensionss.pop()
+        return as_tuple(dimensions), as_tuple(dimensions)
+
+    def __halo_setup__(self, components=(), **kwargs):
+        halos = {i.halo for i in components}
+        if len(halos) > 1:
+            raise ValueError("Components must have the same halo")
+        return halos.pop()
+
+    # Class attributes overrides
+
+    @property
+    def is_DiscreteFunction(self):
+        return self.c0.is_DiscreteFunction
+
+    @property
+    def is_TimeFunction(self):
+        return self.c0.is_TimeFunction
+
+    # Other properties and methods
+
+    @property
+    def components(self):
+        return self._components
+
+    @property
+    def c0(self):
+        # Shortcut for self.components[0]
+        return self.components[0]
+
+    @property
+    def grid(self):
+        if self.is_DiscreteFunction:
+            return self.c0.grid
+        else:
+            return None
+
+    @property
+    def symbolic_shape(self):
+        # A Bundle may be defined over a SteppingDimension, which is of unknown
+        # size, hence we gotta use the actual numeric size instead
+        ret = []
+        for d, s, v in zip(self.dimensions, super().symbolic_shape, self.c0.shape):
+            if d.is_Stepping:
+                ret.append(Number(v))
+            else:
+                ret.append(s)
+        return DimensionTuple(*ret, getters=self.dimensions)
+
+    @property
+    def initvalue(self):
+        return None
+
+    @property
+    def is_shared(self):
+        if self.c0.is_Array:
+            return self.c0._is_shared
+        else:
+            return False
+
+    # CodeSymbol overrides defaulting to self.c0's behaviour
+
+    for i in ['_mem_internal_eager', '_mem_internal_lazy', '_mem_local',
```

Comment:
Why not `getattr` default like in Differentiable or Tensor? Would avoid keeping track of all the function that needs to default to `c0`

Reply:
correct me if I'm wrong, but IIRC `__getattr__` gets called only when an attribute with the given name could not be found. Here the attributes exist, as they are defined in a parent superclass (`CodeSymbol`), so we need explicit overriding... no?

Reply:
Hum now I'm in doubt thought was always called  you probably right